EC-84: Revise Cache, Catalog and Checkout sections (#3)
ezsystems/commerce-documentation: docs/guide/cache/content_cache_refresh/content_cache_refresh.md

# Content cache refresh

The content cache refreshing system enables displaying the current information when objects in the Back Office are modified.

The content cache refresh service is used in the following situations:

- Navigation cache is refreshed for example if one or many navigation nodes are changed
- If a category is changed, the cached names for categories used by the search are updated (`CategoryNameService`)

!!! note

>     You need to have a cronjob set up to use this feature.

Reviewer: ```suggestion
    You need to have a cron job set up to use this feature.
```

> ##  Set up cronjob to refresh caches

Reviewer: ```suggestion
##  Set up cron job to refresh caches
```

> To set up a cronjob to refresh caches, create a system cronjob to run the following console command:

Reviewer: ```suggestion
To set up a cron job to refresh caches, create a system cron job to run the following console command:
```

``` bash
php bin/console silversolutions:cache:refresh
```

## Slot implementation

!!! note

>     See [SignalSlot documentation](https://doc.ezplatform.com/en/2.5/guide/signalslots/#signals-reference) for more information about the Signal Slot system.

Reviewer: ```suggestion
    See [SignalSlot documentation](https://doc.ezplatform.com/en/2.5/guide/signalslots/#signals-reference) for more information about the SignalSlot system.
```

The `ContentModificationSlot` pushes content modification data into the queue by using the `ContentModificationQueueService`.

## Definitions

> Content modification is an array with information about the modified content.

Reviewer: ```suggestion
Content modification is an array of information about the modified content.
```

|Attribute (key)|Value type|Description|
|--- |--- |--- |
|`content_id`|int|Content ID of the modified Content item.|
|`action`|string|Name of the action which resulted in content modification, e.g. `publish` or `delete`.|
|`trigger`|string|Legacy trigger name which caused content modification, e.g. `post_publish`, `pre_delete`. This attribute is created only to provide additional information for content modification processor developers.|

!!! note

    You can find a list of actions and triggers in [supported content modification actions and triggers](#supported-content-modification-actions-and-triggers)

    Content modification processors can provide additional attributes.

Content modification queue is a series of content modifications.
It is implemented with Doctrine.

- `ContentModificationQueue` - Doctrine entity (queue item).
- `ContentModificationQueueRepository` - Doctrine entity repository for queue items.

`ContentModificationQueueService` is a service that stores content modifications in the queue and runs processors to handle content modifications.

> Content modification processor is a service implemented to process information about content modification (e.g. refreshes related caches)

Reviewer: ```suggestion
Content modification processor is a service implemented to process information about content modification (e.g. refreshes related caches).
```

1. Gathers all services that are tagged with `siso_tools.content_modification_processor` (content modification processors).
1. Fetches all current content modifications from the queue.
1. Gets additional attributes from processors.
1. Pushes content modifications to the queue (table `ses_content_modification_queue`).
1. Passes every content modification object to every processor service.
1. Removes processed content modifications from the queue.

![](../../img/content_cache_1.png)

## Content modification processor

Every content modification processor must implement two actions:

1. Get content ID and expose additional attributes that the processor can process later.  
    For example, a navigation cache processor retrieves a Content item's language code from the given content ID. This information is added to content modification data along with initial data.
2. Process content modification data (e.g. clear cache).

### Content modification processor actions

Content modifications are stored in the content modification queue immediately after content is changed.
However, processing of the queue doesn't happen immediately.

> Shop administrator can process the queue manually with a console script or configure a cronjob task.

Reviewer: ```suggestion
Shop administrator can process the queue manually with a console script or configure a cron job task.
```

    $contentTypeService = $this->repository->getContentTypeService();

    // 2. Get data to store in attributes
    $contentType = $contentTypeService->loadContentType($content->contentInfo->contentTypeId);
    $languageCode = $content->getVersionInfo()->initialLanguageCode;

    // 3. Define array of attributes
    return array(
        'content_type' => $contentType->identifier,
        'language_code' => $languageCode,
    );
}
// ...
public function process($queue)
{
    ...

    foreach($queue as $queueItem) {

        $data = $queueItem->getData();

        // Content type attribute is requested
        if (!isset($data['content_type']) || $data['content_type'] !== self::TRANSLATION_IDENTIFIER) {
            continue;
        }

        if(in_array($data['action'], array('delete','removetranslation', 'hide'))) {
            $this->transService->removeTranslationCache(
                ...
                // Language code attribute is requested
                $data['language_code']
            );
        }
    }
}
```

!!! note

    Please pay attention that process method uses attributes `language_code` and `content_type`

>     which were added to content modification array using method `getContentModificationAttributes`.

Reviewer: ```suggestion
    which were added to content modification array using `getContentModificationAttributes` method.
```
